dandi/upload.py

```diff
@@ -573,3 +589,267 @@ def upload_agg(*ignored):
+                if existing == "overwrite":
+                    if extant["sha256"] == sha256_digest:
+                        yield skip_file("file exists")
+                        return
+                elif existing == "refresh":
+                    pass
+                elif existing == "force":
+                    pass
+                else:
+                    raise ValueError("existing")
+
+            #
+            # Validate first, so we do not bother server at all if not kosher
+            #
+            # TODO: enable back validation of dandiset.yaml
+            if path.name != dandiset_metadata_file and validation != "skip":
+                yield {"status": "validating"}
+                validation_errors = validate_file(path)
+                yield {"errors": len(validation_errors)}
+                # TODO: split for dandi, pynwb errors
+                if validation_errors:
+                    if validation == "require":
+                        yield skip_file("failed validation")
+                        return
+                else:
+                    yield {"status": "validated"}
+            else:
+                # yielding empty causes pyout to get stuck or crash
+                # https://github.com/pyout/pyout/issues/91
+                # yield {"errors": '',}
+                pass
+
+            #
+            # Special handling for dandiset.yaml
+            # Yarik hates it but that is life for now. TODO
+            #
+            if path.name == dandiset_metadata_file:
+                # TODO This is a temporary measure to avoid breaking web UI
+                # dandiset metadata schema assumptions.  All edits should happen
+                # online.
+                yield skip_file("should be edited online")
+                return
+
+            #
+            # Extract metadata - delayed since takes time, but is done before
+            # actual upload, so we could skip if this fails
+            #
+            # Extract metadata before actual upload and skip if fails
+            # TODO: allow for for non-nwb files to skip this step
+            # ad-hoc for dandiset.yaml for now
+            yield {"status": "extracting metadata"}
+            try:
+                metadata = nwb2asset(path, digest=sha256_digest, digest_type="sha256")
+            except Exception as exc:
+                if allow_any_path:
+                    yield {"status": "failed to extract metadata"}
+                    metadata = {
+                        "contentSize": os.path.getsize(path),
+                        "digest": sha256_digest,
+                        "digest_type": "sha256",
+                        # "encodingFormat": # TODO
+                    }
+                else:
+                    yield skip_file("failed to extract metadata: %s" % str(exc))
+                    return
+
+            #
+            # Upload file
+            #
+            yield {"status": "uploading"}
```

Comment:
no coverage here suggests that we actually not uploaded any file... so not yet sure how test passes on CI (besides if glob is just wrong) and also for me locally test fails with

```
___________________________________________________ test_new_upload_download ____________________________________________________
/home/yoh/proj/dandi/dandi-cli/dandi/tests/test_upload.py:261: in test_new_upload_download
    upload(
/home/yoh/proj/dandi/dandi-cli/dandi/upload.py:50: in upload
    return _new_upload(
/home/yoh/proj/dandi/dandi-cli/dandi/upload.py:609: in _new_upload
    ds_identifier = dandiset.identifier
/home/yoh/proj/dandi/dandi-cli/dandi/dandiset.py:107: in identifier
    raise ValueError(
E   ValueError: Found no dandiset.identifier in metadata record: {'sex': ['M'], 'number_of_subjects': 1, 'organism': [{'species': 'mouse'}], 'description': 'experiment_description1', 'publications': ['related_publications1']}
```

Reply:
In the course of fixing this, I realized that the code doesn't set the auth token for the new API anywhere.  How should that be passed in/obtained?

Reply:
@dchiquit how to login/get token within the test docker compose instance of api server?

Reply:
@yarikoptic I know how to do *that*; the problem is how to pass the token to the upload function.

Reply:
do we get 401 response if we try to upload without authentication?  if so - guard in `send_request` for receiving 401 and adopt GirderCli's `dandi_authenticate` for RESTFullAPIClient to do authentication in such cases (if was not authenticated before).
to avoid confusion etc, rename prior uses of DANDI_API_KEY to DANDI_GIRDER_API_KEY and use DANDI_API_KEY for authentication against our API server here.

Reply:
The tests & coverage should be working properly now.

Reply:
The auth token can be obtained from the web GUI by clicking on the user menu drop down and copying the `API Key`, same as in girder.

Specifying a `Authorization: Token 123567890abcdefg` header should be all that is required to use that token when making requests.